sklearn/decomposition/tests/test_kernel_pca.py

```diff
@@ -272,3 +276,100 @@ def test_nested_circles():
+         [0, 6 * 1e-5j]]
+    # for solver in solvers:
+    #     kpca = KernelPCA(kernel=kernel_getter, eigen_solver=solver,
+    #                      fit_inverse_transform=False)
+    #     kpca._centerer = IdentityKernelTransformer()
+    #     K = kpca._get_kernel(K)
+    #     with pytest.raises(ValueError):
+    #         # note: we can not test 'fit' because _centerer would be replaced
+    #         kpca._fit_transform(K)
+    #
+    # For safety concerning future evolutions the corresponding code is left in
+    # KernelPCA, and we test it directly by calling the inner method here:
+    with pytest.raises(ValueError):
+        check_kernel_eigenvalues((K[0][0], K[1][1]))
+
+    # All negative eigenvalues: error
+    # -------------------------------
+    K = [[-5, 0],
+         [0, -6e-5]]
+    # check that the inner method works
+    with pytest.raises(ValueError):
+        check_kernel_eigenvalues((K[0][0], K[1][1]))
+
+    for solver in solvers:
+        kpca = KernelPCA(kernel="precomputed", eigen_solver=solver,
+                         fit_inverse_transform=False)
+        kpca._centerer = IdentityKernelTransformer()
+        K = kpca._get_kernel(K)
+        with pytest.raises(ValueError):
+            # note: we can not test 'fit' because _centerer would be replaced
+            kpca._fit_transform(K)
+
+    # Significant negative eigenvalue: warning
+    # ----------------------------------------
+    K = [[5, 0],
+         [0, -6e-5]]
+    # check that the inner method works
+    assert_warns(KernelWarning,
+                 lambda: check_kernel_eigenvalues((K[0][0], K[1][1])))
+
+    for solver in solvers_except_arpack:
+        # Note: arpack detects this case and raises an error already
+        kpca = KernelPCA(kernel="precomputed", eigen_solver=solver,
+                         fit_inverse_transform=False)
+        kpca._centerer = IdentityKernelTransformer()
+        K = kpca._get_kernel(K)
+        # note: we can not test 'fit' because _centerer would be replaced
+        assert_warns(KernelWarning, lambda: kpca._fit_transform(K))
+
+    # Bad conditioning
+    # ----------------
+    K = [[5, 0],
+         [0, 4e-12]]
+    # check that the inner method works
+    assert_warns(KernelWarning,
+                 lambda: check_kernel_eigenvalues((K[0][0], K[1][1]),
+                                                  warn_on_zeros=True))
+
+    # This is actually a normal behaviour when the number of samples is big,
+    # so no special warning should be raised in this case
+    for solver in solvers_except_arpack:
+        # Note: arpack detects this case and raises an error already
+        kpca = KernelPCA(kernel="precomputed", eigen_solver=solver,
+                         fit_inverse_transform=False)
+        kpca._centerer = IdentityKernelTransformer()
+        K = kpca._get_kernel(K)
+        # note: we can not test 'fit' because _centerer would be replaced
+        assert_no_warnings(lambda: kpca._fit_transform(K))
```

Comment:
Please use `with pytest.warns(None) as w: ... assert not w`

Reply:
ok.